File: src/프로그래머스/unrated/169199. 리코쳇 로봇/리코쳇 로봇.py
from collections import deque

dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]


def solution(board):
    robot = None
    board = [list(x) for x in board]
    x_len = len(board)
    y_len = len(board[0])
    for x in range(x_len):
        for y in range(y_len):
            if board[x][y] == "R":
                robot = [x, y]
                board[x][y] = "."

    visited = [[False for _ in range(y_len)] for _ in range(x_len)]
    visited[robot[0]][robot[1]] = True
    q = deque([[0, robot[0], robot[1]]])
    while q:
        count, x, y = q.popleft()
        for k in range(4):
            xx = x
            yy = y

            while 0 <= xx < x_len and 0 <= yy < y_len and board[xx][yy] != "D":
                xx += dx[k]
                yy += dy[k]
            else:
                xx -= dx[k]
                yy -= dy[k]
                
            if 0 <= xx < x_len and 0 <= yy < y_len:
                if board[xx][yy] == "G":
                    return count + 1
                if not visited[xx][yy] and board[xx][yy] == ".":
                    visited[xx][yy] = True
                    q.append([count + 1, xx, yy])
    return -1

File: src/프로그래머스/unrated/169199. 리코쳇 로봇/리코쳇 로봇.java
import java.util.ArrayDeque;

class Solution {
    private final int[] dx = { 0, 0, 1, -1 };
    private final int[] dy = { 1, -1, 0, 0 };
    private int xLen;
    private int yLen;

    public int solution(String[] board) {
        xLen = board.length;
        yLen = board[0].length();
        int[][] visited = new int[xLen][yLen];
        ArrayDeque<Vo> queue = new ArrayDeque<>();

        for (int i = 0; i < xLen; i++) {
            for (int j = 0; j < yLen; j++) {
                if (board[i].charAt(j) == 'R') {
                    queue.addLast(new Vo(0, i, j));
                    visited[i][j] = 1;
                    board[i] = board[i].replace('R', '.');
                }
            }
        }

        while (!queue.isEmpty()) {
            Vo vo = queue.removeFirst();
            int count = vo.count;
            for (int k = 0; k < 4; k++) {
                int xx = vo.x;
                int yy = vo.y;

                while (checkLocation(xx, yy) && board[xx].charAt(yy) != 'D') {
                    xx += dx[k];
                    yy += dy[k];
                }

                if (!checkLocation(xx, yy) || board[xx].charAt(yy) == 'D') {
                    xx -= dx[k];
                    yy -= dy[k];
                }

                if (checkLocation(xx, yy)) {
                    if (board[xx].charAt(yy) == 'G') {
                        return count + 1;
                    }
                    if (visited[xx][yy] != 1 && board[xx].charAt(yy) == '.') {
                        visited[xx][yy] = 1;
                        queue.addLast(new Vo(count + 1, xx, yy));
                    }
                }

            }
        }

        return -1;
    }

    private boolean checkLocation(int x, int y) {
        return (x >= 0 && x < xLen) && (y >= 0 && y < yLen);
    }

    class Vo {
        public Vo(int count, int x, int y) {
            this.count = count;
            this.x = x;
            this.y = y;
        }

        private int count;
        private int x;
        private int y;

    }
}
